### PR TITLE
Refactoring folder structure, code formatting and docker files

### PR DESCRIPTION
**Pull request by piteur**
What changed:
 - added `.editorconfig` file and enforced it on existing files
 - `Dockerfile` using multi-stage for reduced container size
 - moved `frontend` resources to a dedicated subfolder

# Artifacts

## Dockler images

Docker image before: `biosmarcel/scribble.rs   latest              8dd349e2623d        2 days ago          831MB`
Docker image now: `biosmarcel/scribble.rs   latest              2083610e5296        5 minutes ago       10.8MB`

We lost around **820Mb**.

## Change port on docker container

Docker container now use `entrypoint` which make us able to change parameters on run:
```shell
docker run -p 8081:8081 2083610e5296 --portHTTP=8081
2020/04/10 00:53:17 Started
[...]
docker run -p 8080:8080 2083610e5296 --portHTTP=8080
2020/04/10 00:53:17 Started
```

### CODE REVIEW

**Comment by piteur** [comment]
Messed up a bit the rebase, cleaned-up.

**Comment by Bios-Marcel** [comment]
Regarding editor config. I dislike that. The go way would be to enforce execution of `gofmt -s -w` on all files.

File: Dockerfile [hunk 1]
```diff
@@ -1,7 +1,23 @@
...
+
+# certificates
+FROM alpine:latest as certs
+RUN apk --no-cache add ca-certificates
+
+# runner
+FROM scratch
+#WORKDIR /app
+
+COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
```
[the remark this answers is not in the document]
Comment (piteur) [reply Dockerfile:19]: Always a pleasure to explain!

This is not required right now. But I did it, in case of future features.
The goal here is to have the smallest container possible, as only the binary file is required. When you use a `scratch` container, it's an empty one: **0Kib**.

But, when you want to do an API call from the Golang app, you mostly need to do HTTPS calls. And in order to do that, you need to have the certificate authority to validate the SSL certificates. Without that file, no HTTPS calls are possible.

Here, I simply use the one from the `alpine:latest` to put the file in the scratch container.

This is a common best practice in Golang containers.

Hope that helps :)

TLDR:
 - use stage container to get the CA file
 - copy it to the scratch container beside the binary

File: templates/lobby.html [hunk 2]
```diff
@@ -426,9 +426,7 @@
         } else if (parsed.type === "clear-drawing-board") {
             clear(context)
         } else if (parsed.type === "next-turn") {
-            //If a player doesn't choose, the dialog will still be up.
-            wordDialog.style.visibility = "hidden";
-            playWav('/resources/end-turn.wav');
```
[the remark this answers is not in the document]
Comment (piteur) [reply templates/lobby.html:431]: The reason is that I fucked up my rebase later this night :D
Will fix that mess,

File: resources/frontend/style.css [hunk 3]
```diff
@@ -1,13 +1,13 @@
 @font-face {
     font-family: 'Montserrat';
     font-weight: normal;
-    src: local('Monserrat'), url(/resources/Montserrat-Regular.otf) format("opentype");
+    src: local('Monserrat'), url(/resources/frontend/Montserrat-Regular.otf) format("opentype");
```
[the remark this answers is not in the document]
Comment (piteur) [reply resources/frontend/style.css:4]: Done.

**Comment by piteur** [comment]
@Bios-Marcel 
> Regarding editor config. I dislike that. The go way would be to enforce execution of gofmt -s -w on all files. If you are using Goland, VS-Code and Vim for example, that pretty much happens already.

I agree and disagree at the same time.
Yes Golang enforce it by default which is bullet-proof for most contribution.
But in the order hand, other settings in the `.editorconfig` file applies to the `*.go` files (charset, eol, whitespace etc...). For me it made sense to enforce it, **explicitely**.

Up to you, can remove it, just giving my two cents ^^

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (Bios-Marcel) [reply Dockerfile:19]: Would be great if you could put the short version of that comment into the docker file itself. Even though it might seem verbose or useless, it will surely help future readers of the code.

**Comment by Bios-Marcel** [comment]
Well, if we want to be strict, i'd even go as far as formatting all code on commit / PR and compare it to it's pre-formatted state. If there's any change, we simply error. Could put this into the wiki / readme. We might as well write a proper contribution guide, whatever else we might have to put in there. But I'd like to go for the most idiomatic way here. I don't know how other projects solve this and haven't ever looked into that. I tried using editorconfig in the past and it wasn't too useful, as it only  worked half-way in most languages / editors. Invoking gofmt however is fully editor agnostic, even without the editor having to do anything for it.

**Comment by piteur** [comment]
Removed the go specification from the `.editorconfig` file.
Addressed MR comments. Good for another review :)

File: README.md [hunk 4]
```diff
@@ -28,8 +28,7 @@ Run the following to build the application:
 ```shell
 git clone https://github.com/scribble-rs/scribble.rs.git
 cd scribble.rs
-go run github.com/markbates/pkger/cmd/pkger -include /resources -include /templates
-go build -o scribblers .
+make build
```
Comment (Bios-Marcel) [inline README.md:31]: That's a nope from me. This won't run on windows ;)

Maybe make a second section in a "show more"-tag that shows the less general solution usable on systems with make.

File: Dockerfile [hunk 5]
```diff
@@ -1,7 +1,24 @@
...
+FROM golang:latest AS builder
 RUN mkdir /app
 ADD . /app/
 WORKDIR /app
+
 RUN go mod download
-RUN go build -o main .
-CMD ["/app/main", "--portHTTP=80"]
\ No newline at end of file
+RUN make build
```
Comment (Bios-Marcel) [inline Dockerfile:8]: Not sure how docker on windows works, but will this command run inside of a container or no the host system?
I wanna keep everything as compatible as possible.

Comment (piteur) [reply Dockerfile:8]: The command will be run inside the docker container, like when you're running an `apt get whatever`.

File: Dockerfile [hunk 6]
```diff
@@ -1,7 +1,24 @@
...
+FROM alpine:latest as certs
+RUN apk --no-cache add ca-certificates
+
+# runner
+FROM scratch
+#WORKDIR /app
+
+COPY --from=certs /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
+COPY --from=builder /app/scribblers /scribblers
+COPY --from=builder /app/templates /templates/
```
Comment (Bios-Marcel) [inline Dockerfile:22]: Uhm, since we invoke `make` which invokes `pkger`, i believe this isn't necessary anymore, as it simply duplicates the templates? They are part of the binary either way if `pkger` was invoked.

Comment (piteur) [reply Dockerfile:22]: The files are not part of the binary, as the binary explicitly calls the filesystem to load them.
If you don't add the files, the binary panic with that error message:

```
$ docker run 92b389cabc41
panic: open /templates/error.html: no such file or directory

goroutine 1 [running]:
github.com/scribble-rs/scribble.rs/communication.readTemplateFile(0x87be83, 0xa, 0x0, 0x0)
	/app/communication/init.go:73 +0x1c8
github.com/scribble-rs/scribble.rs/communication.init.0()
	/app/communication/init.go:23 +0x5d
```

Comment (Bios-Marcel) [reply Dockerfile:22]: Hm, hold on. This shouldn't happen. This is exactly what pkger does. pkger represents a virtual file-system that falls back to the real file system if the files aren't packaged.

Comment (piteur) [reply Dockerfile:22]: Ok found the issue: templates are not loaded with `pkger`:
```go
errorPage, parseError = template.New("error.html").Parse(readTemplateFile("error.html"))
```
Compared to templates:
```go
templateHandle, pkgerError := pkger.Open("/templates/" + name)
```

Comment (Bios-Marcel) [reply Dockerfile:22]: Appears that smth is wrong there. I'll look into it.

File: README.md (hunk 4 above)
Comment (piteur) [reply README.md:34]: Added more details there.

I'm wondering: when developing on Windows nowadays, people still use Powershell in favor of Linux Subsystem? It has been ages without Windows as a dev env for me.

File: Dockerfile (hunk 6 above)
Comment (piteur) [reply Dockerfile:22]: Do you believe this should be fixed in this MR?
Or the fix coming in will update the Dockerfile?

Comment (Bios-Marcel) [reply Dockerfile:20]: I am not sure what the problem is yet, so I don't know. The documentation on pkger is either not sufficient enough, there's a bug or I am stupid. One of those three, not sure which one it is yet :D

Comment (Bios-Marcel) [reply Dockerfile:20]: Well, I created an issue. Let's just wait til we resolve this: https://github.com/markbates/pkger/issues/86

If nobody replies, we'll apply the workaround I presented. Alternatively we don't really need to execute packager but can simply copy the resources and templates folder into the docker container, right?

Comment (piteur) [reply Dockerfile:20]: Will add a workaround comment for the Dockerfile line, with link to the issue.
And we can move forward with this MR? Will like to take a look at this missbahvior as well.

Comment (piteur) [reply Dockerfile:21]: Comment added.

File: README.md (hunk 4 above)
Comment (Bios-Marcel) [reply README.md:34]: Oh, i intended it to be a new codeblock, that way the lazy bums (like me), can just copy the whole block ^^

sorry for the bickering :smile:

Comment (piteur) [reply README.md:36]: Fixed, I'm used to be nitpicking king too, no worries :)

**Comment by Bios-Marcel** [comment]
I'll apply the workaround I mentioned. It's a bit annoying, but no need to stifle on it :)
